### DISCUSSION TITLE
[mini.pick] added `mini.pick` as a picker for `zk-nvim`

### DISCUSSION THREAD

**Discussion by pkazmier** [nvim-mini/mini.nvim#785] [Show and tell]
I use [zk](https://github.com/zk-org/zk) and its neovim plug-in [zk-nvim](https://github.com/zk-org/zk-nvim) extensively. This is a markdown-based note taking system with fantastic integration with neovim (LSP, support for various pickers, etc ...). Out of the box, `zk-nvim` supports the following pickers: `vim.ui.select`, `fzf`, `fzf_lua`, and `telescope`. 

Over the past several weekends, I've been working on an almost exclusive mini configuration (sorry, `mini.pairs`).  I wanted to see if I could add support for `mini.pick`, so I could eliminate telescope altogether. I'm happy to say it was fairly trivial (well, it took me a few hours, mostly reading docs).

Here's a quick demo which I suppose will just look like  any other file-based source. But, nonetheless, I thought I'd share in case others use `zk-nvim`. I'll be submitting a PR to them tomorrow.

https://github.com/echasnovski/mini.nvim/assets/747855/4a90bdeb-bf75-45ea-90c2-51cb0a99af4e

p.s. Not sure how to make good screen recordings like @echasnovski. I'm on a Mac and this was just a QuickTime screen recording. You can't see keys I'm typing unfortunately, but the first sequence was `<leader>ns` (+Notes -> Search), and the second was `<leader>nt` (+Notes -> Tags).

Here's how I integrated it. The three functions here must be implemented for a `zk-nvim` picker, so it was easy to plug-in in `mini.pick`. I just had to wrap the note object in a new table for the `text` and `path` fields to take advantage of the builtins in `mini.pick`. Same with the tag object in the tag picker, but this time so I could format it nicely (count of tags with the tag name). In both cases, I had to use my own choose that just returns `true` to stop the picker because `zk-nvim` has its own callback where it opens the notes.

```lua
local M = {}
local H = {}
local minipick = require("mini.pick")

-- Picker from zk-nvim must implement this.
M.note_picker_list_api_selection = { "title", "path", "absPath" }

-- Picker from zk-nvim must implement this.
M.show_note_picker = function(notes, opts, cb)
  notes = vim.tbl_map(function(n)
    return { text = n.title, path = n.absPath, note = n }
  end, notes)

  opts = opts or {}
  local minipick_opts = vim.tbl_deep_extend("force", {
    source = {
      items = notes,
      name = opts.title,
      choose = function(_)
        return nil
      end,
    },
    window = {
      prompt_prefix = opts.title .. "> ",
    },
  }, opts.minipick or {})

  local item = minipick.start(minipick_opts)
  if item ~= nil then
    cb(opts.multi_select and { item.note } or item.note)
  end
end

-- Picker from zk-nvim must implement this.
M.show_tag_picker = function(tags, opts, cb)
  tags = vim.tbl_map(function(t)
    local padded_cnt = H.ensure_text_width(t.note_count, opts.note_count_width or 4)
    return {
      text = string.format("%s │ %s", padded_cnt, t.name),
      tag = t,
    }
  end, tags)

  opts = opts or {}
  local minipick_opts = vim.tbl_deep_extend("force", {
    source = {
      items = tags,
      name = opts.title,
      choose = function(_)
        return nil
      end,
    },
    window = {
      prompt_prefix = opts.title .. "> ",
    },
  }, opts.minipick or {})

  local item = minipick.start(minipick_opts)
  if item ~= nil then
    cb(opts.multi_select and { item.tag } or item.tag)
  end
end

-- From mini.extras
H.ensure_text_width = function(text, width)
  local text_width = vim.fn.strchars(text)
  if text_width <= width then
    return string.rep(" ", width - text_width) .. text
  end
  return "…" .. vim.fn.strcharpart(text, text_width - width + 1, width - 1)
end

return M

```

**Comment by echasnovski** [comment]
Another one great customization! Thanks for sharing!

My intuition says that using overridden `vim.ui.select()` (with [ui_select()](https://github.com/echasnovski/mini.nvim/blob/d0b297ab233a57bed00ab69bd66135dee111967f/doc/mini-pick.txt#L1025)) should be enough for most cases. Although it won't support small quality of life improvements like preview, etc.

(Hope all 1:1 meetings went well)

> p.s. Not sure how to make good screen recordings like @echasnovski. I'm on a Mac and this was just a QuickTime screen recording. You can't see keys I'm typing unfortunately, but the first sequence was `<leader>ns` (+Notes -> Search), and the second was `<leader>nt` (+Notes -> Tags).

Those screen recordings are done on Linux with [SimpleScreenRecorder](https://github.com/MaartenBaert/ssr) and [Screenkey](https://www.thregr.org/wavexx/software/screenkey/).

**Comment by pkazmier** [comment]
> (Hope all 1:1 meetings went well)

LOL. In my day job, since July of 2022 when I switched to neovim from doom emacs (and org mode), I've accumulated 1,205 notes in `zk`—of which 730 of those are 1:1s with my associates. The majority of my time in neovim is spent taking notes. So, I've optimized this workflow with custom `zk-nvim` commands, which are now able to use `mini.pick`. Some of them dynamically set mappings, which get passed through to the integration above.

> My intuition says that using overridden vim.ui.select() (with [ui_select()](https://github.com/echasnovski/mini.nvim/blob/d0b297ab233a57bed00ab69bd66135dee111967f/doc/mini-pick.txt#L1025)) should be enough for most cases. Although it won't support small quality of life improvements like preview, etc.

Yeah, in this case, the small quality of life improvements add up quickly given how much time I spend taking notes, so I wanted a full integration. Plus, I also wanted to help spread the word about `mini.pick` as a telescope alternative by contributing another picker to `zk-nvim` to sit alongside this comment in the `README`:
```
require("zk").setup({
  -- can be "telescope", "fzf", "fzf_lua" or "select" (`vim.ui.select`)
  -- it's recommended to use "telescope", "fzf" or "fzf_lua"
  picker = "select",
```

**Reply by echasnovski** [reply]
> Plus, I also wanted to help spread the word about `mini.pick` as a telescope alternative by contributing another picker to `zk-nvim` to sit alongside this comment in the `README`:

Huge thanks for this! Currently it is indeed one of the best ways to contribute to 'mini.nvim' without having to deal with its code.

**Reply by pkazmier** [reply]
FYI. My PR to zk-nvim that adds support for `mini.pick` has been merged into main. 

https://github.com/zk-org/zk-nvim/commit/e2b6d62b18a88249016bf917d4e5bb0e417ac974
